File: LeetCode/124.Binary_Tree_Maximum_Path_Sum/Mario/124.Binary_Tree_Maximum_Path_Sum.py
#!/usr/bin/env python3

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
	def maxPathSum(self, root: Optional[TreeNode]) -> int:
		self.res = root.val
		
		def dfs(node):
			if not node:
				return 0
			
			rightMax = dfs(node.right)
			leftMax = dfs(node.left)
			leftMax = max(leftMax, 0)
			rightMax = max(rightMax, 0)
			
			self.res = max(self.res, node.val+leftMax+rightMax)
			
			return node.val + max(leftMax, rightMax)
		
		dfs(root)
		return self.res

File: LeetCode/124.Binary_Tree_Maximum_Path_Sum/124.Binary_Tree_Maximum_Path_Sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {

    private int maxValue = -1001;

    public int maxPathSum(TreeNode root) {
        dfs(root);
        return maxValue;
    }

    private int dfs(TreeNode node) {
        if (node == null) {
            return 0;
        }

        int leftMax = dfs(node.left);
        int rightMax = dfs(node.right);
        leftMax = Math.max(leftMax, 0);
        rightMax = Math.max(rightMax, 0);

        maxValue = Math.max(maxValue, leftMax + rightMax + node.val);

        return Math.max(leftMax, rightMax) + node.val;
    }
}